Implement sidebarAction/sidePanel open/close/toggle API methods.

AriYoung00 opened the pull request:
#### be5aff9f77c6050e47b364c7d2db9fa73d197df4
<pre>
Implement sidebarAction/sidePanel open/close/toggle API methods.
<a href="https://webkit.org/b/279684">https://webkit.org/b/279684</a>
<a href="https://rdar.apple.com/135894111">rdar://135894111</a>

Reviewed by Timothy Hatcher.

This PR fills in the implementation of the `sidebarAction.open()`,
`sidebarAction.close()`, `sidebarAction.toggle()`, and
`sidePanel.open()` APIs. It also fixes a bug in the implementation
of `sidebarAction.isOpen`. Finally, this PR also adds a reasonably
comprehensive set of unit tests that exercise the behavior of these
API methods.

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension _hasSidebar]): Replace WebExtension::hasSidebar with
WebExtension::hasAnySidebar
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::toExpected): Replace call to toWebExtensionError with
makeUnexpected
(WebKit::getSidebarWithIdentifiers): Replace calls to
toWebExtensionError with makeUnexpected
(WebKit::getOrCreateSidebarWithIdentifiers): Replace calls to
toWebExtensionError with makeUnexpected
(WebKit::scopedAPINameFor): Add helper method to create full API name
for methods which service both sidebarAction and sidePanel
(WebKit::WebExtensionContext::openSidebar): Replaced openSidebarForTab
with openSidebar (allows for less-convoluted code in other locations).
Removed call to fireActionClickEventIfNeeded, added call to delegate to
actually open sidebar.
(WebKit::WebExtensionContext::closeSidebar): Replaced closeSidebarForTab
with closeSidebar, added call to delegate to actually close sidebar.
(WebKit::WebExtensionContext::canProgrammaticallyOpenSidebar): Moved
this method to WebExtensionContext rather than WebExtensionSidebar,
added implementation.
(WebKit::WebExtensionContext::canProgrammaticallyCloseSidebar): Moved
this method to WebExtensionContext rather than WebExtensionSidebar,
added implementation.
(WebKit::WebExtensionContext::sidebarOpen): Simplified implementation
based on other changes in this PR, added condition that relevant tab
must have active user gesture to programmatically open sidebar.
(WebKit::WebExtensionContext::sidebarClose): Cleaned up implementation,
added condition that relevant tab must have active user gesture to
programmatically close sidebar.
(WebKit::WebExtensionContext::sidebarIsOpen): Removed code which checks
if window-specific sidebar is open (window-specific objects should never
directly be open).
(WebKit::WebExtensionContext::sidebarToggle): Fixed minor issues in
implementation, cleaned up implementation.
(WebKit::WebExtensionContext::sidebarGetTitle): Add API name to error
message
(WebKit::WebExtensionContext::sidebarSetTitle): Add API name to error
message
(WebKit::WebExtensionContext::sidebarGetOptions): Add API name to error
message
(WebKit::WebExtensionContext::sidebarSetOptions): Add API name to error
message
(WebKit::WebExtensionContext::openSidebarForTab): Deleted (renamed to
openSidebar).
(WebKit::WebExtensionContext::closeSidebarForTab): Deleted (renamed to
closeSidebar).
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasSidebarAction): Helper method to check whether
this extension uses sidebarAction
(WebKit::WebExtension::hasSidePanel): Helper method to check whether
this extension uses sidePanel
(WebKit::WebExtension::hasAnySidebar): Helper method to check whether
this extension uses sidebarAction or sidePanel
(WebKit::WebExtension::hasSidebar): Deleted (replaced with
hasAnySidebar).
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getOrCreateSidebar): Replace calls to
WebExtension::hasSidebar with WebExtension::hasAnySidebar
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(WebKit::WebExtensionSidebar::canProgrammaticallyOpenSidebar const): Deleted
(moved to WebExtensionContext).
(WebKit::WebExtensionSidebar::openSidebarWhenReady): Deleted.
(WebKit::WebExtensionSidebar::canProgrammaticallyCloseSidebar const): Deleted
(moved to WebExtensionContext).
(WebKit::WebExtensionSidebar::closeSidebarWhenReady): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Add declarations of
  hasSidebarAction, hasSidePanel, hasAnySidebar; remove declaration of
  hasSidebar
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Removed
  declarations of openSidebarForTab and closeSidebarForTab, added
declarations of openSidebar, closeSidebar,
canProgrammaticallyOpenSidebar, and canProgrammaticallyCloseSidebar
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Removed
  declarations of canProgrammaticallyOpenSidebar, openSidebarWhenReady,
canProgrammaticallyCloseSidebar, and closeSidebarWhenReady.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
(WebKit::parseWindowIdentifier): Fix bug (accidentally typo&apos;d tabIdKey
rather than windowIdKey).
(WebKit::WebExtensionAPISidePanel::open): Add user gesture check
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::WebExtensionAPISidebarAction::open): Add user gesture check
(WebKit::WebExtensionAPISidebarAction::close): Add user gesture check
(WebKit::WebExtensionAPISidebarAction::toggle): Add user gesture check
(WebKit::WebExtensionAPISidebarAction::isOpen): Only return error for
invalid window ID if windowId has a value AND isValid(windowId) also
fails, rather than just if isValid(windowId) fails
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm: Add
  import for _WKWebExtensionSidebar.h, add default path to
sidePanelManifest global
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionClearTabPanel)): Renamed old test to fix typo
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionOpenFailsWithoutUserGesture)):
Added test ensuring that sidebarAction.open() fails if the current tab
has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionCloseFailsWithoutUserGesture)):
Added test ensuring that sidebarAction.close() fails if the current tab
has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionToggleFailsWithoutUserGesture)):
Added test ensuring that sidebarAction.toggle() fails if the current tab
has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionOpenSucceedsWithUserGesture)):
Added test ensuring that sidebarAction.open() succeeds if the current
tab has a user gesture. Also ensures that the relevant delegate method
is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionCloseSucceedsWithUserGesture)):
Added test ensuring that sidebarAction.close() succeeds if the current
tab has a user gesture. Also ensures that the relevant delegate method
is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionToggleSucceedsWithUserGesture)):
Added test ensuring that sidebarAction.toggle() succeeds if the current
tab has a user gesture. Also ensures that the relevant delegate method
is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpenForTabFailsWithoutUserGesture)):
Added test ensuring that sidePanel.open({ tabId: ... }) fails if the
current tab has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpenForWindowFailsWithoutUserGesture)):
Added test ensuring that sidePanel.open({ windowId: ... }) fails if the
current tab has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtnsionAPISidebar, SidePanelOpenForTabSucceedsWithUserGesture)):
Added test ensuring that sidePanel.open({ tabId: ... }) succeeds if the
current tab has a user gesture. Also ensures that the relevant delegate
method is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpenForWindowSucceedsWithUserGesture)):
Added test ensuring that sidePanel.open({ windowId: ... }) succeeds if
the current tab has a user gesture. Also ensures that the relevant
delegate method is called correctly, and that the sidebar is opened in
the correct window.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarAcionClearTabPanel)): Deleted (renamed to SidebarActionClearTabPanel).
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h: Added
  presentSidebar, closeSidebar, and didUpdateSidebar callback
properties.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:presentSidebar:forExtensionContext:completionHandler:]):
Added.
(-[TestWebExtensionsDelegate _webExtensionController:closeSidebar:forExtensionContext:completionHandler:]):
Added.
(-[TestWebExtensionsDelegate _webExtensionController:didUpdateSidebar:forExtensionContext:]):
Added.

Canonical link: <a href="https://commits.webkit.org/283934@main">https://commits.webkit.org/283934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b995be748702ecd4949a69726a8b3a1b060a279f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47025 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71697 "Built successfully") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54824 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18587 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54159 "Passed tests") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70710 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43147 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58499 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34628 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15904 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17142 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61782 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16246 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73395 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11606 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61605 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11641 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58567 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61655 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9473 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3090 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10316 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42828 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43906 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45095 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43647 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -90,59 +91,89 @@
-            }
-
-            sidebar = WTFMove(maybeSidebar.value());
+        if (auto window = frontmostWindow())
```
nit: `RefPtr window = frontmostWindow()`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -176,19 +206,24 @@
     }
 
+    if (!hasActiveUserGesture(*tab)) {
+        completionHandler(toWebExtensionError(nil, nil, @"open can only be called in response to a user action"));
```
This is the wrong user gesture to check. The check needs to happen in the web process.

Do that via:
```
WebCore::UserGestureIndicator::processingUserGesture()
```

See `WebExtensionAPIPermissions::request()` for an example.

AriYoung00 replied:
Ahh I was wondering if this was the case, thank you!

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -176,19 +206,24 @@
     }
 
+    if (!hasActiveUserGesture(*tab)) {
+        completionHandler(toWebExtensionError(nil, nil, @"open can only be called in response to a user action"));
```
i would spell out the full api name here as opposed to just "open". i think you could pass it in when you call this method from the WebProcess

AriYoung00 replied:
Unfortunately since this services both `sidebarAction.open()` and `sidePanel.open()` it's hard to know which one is in use here. I was thinking about maybe trying to augment the error message on the WebProcess side, since it feels a tad excessive to me to push the API name through IPC just for error message generation.

xeenon replied:
In this case you would put the function name in the first argument of `toWebExtensionError()`. We should do that! But in this case, the error would be in the web process with the correct gesture check so it does not matter here.

For the API name, you can add a helper that returns the right "sidePanel" or "sidebarAction" string based on the manifest entries.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -176,19 +206,24 @@
+
     if (!sidebar) {
-        completionHandler(makeUnexpected(sidebar.error()));
+        completionHandler(toWebExtensionError(nil, nil, @"could not get sidebar for active tab"));
```
"active tab" is a bit of a loaded term in Web Extensions. Maybe just "tab"? Also keep in mind that "because " is prefixed to these errors, so this should be more like "a sidebar should not be found for the tab" so it flows with the full message.

Also, is this error actionable by the developer or is it more of an "unknown error"? If the later, then consider a more generic error. "an unknown error occurred".

xeenon replied:
Also add API name for first parameter.

AriYoung00 replied:
That's a good point, I don't think there's much the developer could do about this error -- I'll change it to "unknown..."

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -176,19 +206,24 @@
+    }
+
+    std::optional<Ref<WebExtensionSidebar>> sidebar = getOrCreateSidebar(*tab);
+
```
Remove line here.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -207,20 +242,24 @@
 
-    auto maybeSidebar = getOrCreateSidebar(*window);
+    if (!hasActiveUserGesture(*tab)) {
+        completionHandler(toWebExtensionError(nil, nil, @"close can only be called in response to a user action"));
```
ditto

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -207,20 +242,24 @@
+    if (!canProgrammaticallyCloseSidebar()) {
         completionHandler(toWebExtensionError(nil, nil, @"it is not implemented"));
         return;
     }
```
Should this be earlier, so it fails faster?

xeenon replied:
Also add API name for first parameter.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -266,29 +302,28 @@
+    if (!canProgrammaticallyCloseSidebar() || !canProgrammaticallyOpenSidebar()) {
+        completionHandler(toWebExtensionError(nil, nil, @"it is not implemented"));
+        return;
     }
```
Should this be earlier, so it fails faster?

xeenon replied:
Also add API name for first parameter.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -149,7 +149,7 @@ static ParseResult parseSidebarActionDetails(NSDictionary *details)
 
     std::optional<WebExtensionWindowIdentifier> windowId = maybeWindowId ? toWebExtensionWindowIdentifier(((NSNumber *) maybeWindowId).doubleValue) : std::nullopt;
-    if (!isValid(windowId)) {
+    if (windowId.has_value() && !isValid(windowId)) {
```
```suggestion
    if (windowId && !isValid(windowId)) {
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -714,6 +719,204 @@
+    auto *script = @[
+        @"await browser.sidebarAction.open()",
+        @"    .then(() => browser.test.notifyFail('sidebarAction.open() did not throw without user gesture'))",
+        @"    .catch(() => browser.test.notifyPass())",
```
You don't `await` with `.then()` or `.catch()`. You would use `try { } catch(e) { }` like a normal sync call.

Instead, use:

```suggestion
        @"browser.test.assertThrows(() => browser.sidebarAction.open())",
        @"browser.test.notifyPass()",
```

The assert will fail the test, even if the pass is hit later.

AriYoung00 replied:
Don't `.then` and `.catch` themselves return promises chained against the initial promise? My intention here is to await the final promise returned by `.catch`.  Regarding `assertThrows` -- that's actually what I tried to use initially. I found that the error returned from `open()` in this scenario only surfaces if the promise returned by `open()` is awaited, so:
```
browser.test.assertThrows(() => browser.sidebarAction.open())
```
 will fail even if the promise is correctly rejected. Unfortunately, we can't do this:
```
browser.test.assertThrows(() => await browser.sidebarAction.open())
```
since it's a syntax error to await in a non-async closure. I also tried this:
```
browser.test.assertThrows(async () => await browser.sidebarAction.open())
```
which doesn't work since `assertThrows` will not await the promise returned by the async closure. So far as I can tell, I don't think we can use `assertThrows` here since the promise is rejected, rather than an exception being directly thrown. This is how I came to the `.then() ... .catch()` chaining. For what it's worth, I played with it experimentally a bit and found that it correctly passes/fails the test based on whether the promise resolves or is rejected.

xeenon replied:
Ah yes! Makes sense. However, the gesture error should be a thrown exception from the web process and not an async error. So `assertThrows` would work then.

AriYoung00 replied:
Fair enough! I forgot about that incoming change as I was writing all this out haha

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -714,6 +719,204 @@
+    auto *script = @[
+        @"await browser.sidebarAction.close()",
+        @"    .then(() => browser.test.notifyFail('sidebarAction.close() did not throw without user gesture'))",
+        @"    .catch(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -714,6 +719,204 @@
+    auto *script = @[
+        @"await browser.sidebarAction.toggle()",
+        @"    .then(() => browser.test.notifyFail('sidebarAction.close() did not throw without user gesture'))",
+        @"    .catch(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -714,6 +719,204 @@
+        @"browser.test.yield('Apply user gesture')",
+        @"await browser.sidebarAction.open()",
+        @"    .catch(() => browser.test.notifyFail('sidebarAction.open() threw with user gesture'))",
+        @"    .then(() => browser.test.notifyPass())",
```
This needs to use `Util::runScriptWithUserGesture()`. See `TEST(WKWebExtensionAPIPermissions, AcceptPermissionsDenyMatchPatternsRequest)` for an example.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -714,6 +719,204 @@
+        @"    .then(() => browser.test.notifyPass())",
+        @"await browser.sidebarAction.close()",
+        @"    .catch(() => browser.test.notifyFail('sidebarAction.close() threw with user gesture'))",
+        @"    .then(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -714,6 +719,204 @@
+        @"browser.test.yield('Apply user gesture')",
+        @"await browser.sidebarAction.toggle()",
+        @"    .catch(() => browser.test.notifyFail('sidebarAction.toggle() threw with user gesture'))",
+        @"    .then(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -948,6 +1151,147 @@
+{
+    auto *script = @[
+        @"let tabs = await browser.tabs.query({})",
+        @"await browser.sidePanel.open({ tabId: tabs[0].id })",
+        @"    .then(() => browser.test.notifyFail('sidePanel.open() did not throw without user gesture'))",
+        @"    .catch(() => browser.test.notifyPass())",
```
Ditto on `await`.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -948,6 +1151,147 @@
+{
+    auto *script = @[
+        @"let window = await browser.windows.getCurrent()",
+        @"await browser.sidePanel.open({ windowId: window.id })",
+        @"    .then(() => browser.test.notifyFail('sidePanel.open() did not throw without user gesture'))",
+        @"    .catch(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -948,6 +1151,147 @@
+        @"let tabs = await browser.tabs.query({})",
+        @"await browser.sidePanel.open({ tabId: tabs[0].id })",
+        @"  .catch(() => browser.test.notifyFail('sidePanel.open() threw with user gesture'))",
+        @"  .then(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -948,6 +1151,147 @@
+        @"let window = await browser.windows.getCurrent()",
+        @"await browser.sidePanel.open({ windowId: window.id })",
+        @"    .catch(() => browser.test.notifyFail('sidePanel.open() threw with user gesture'))",
+        @"    .then(() => browser.test.notifyPass())",
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,34 +32,37 @@
 {
     if (windowIdentifier && tabIdentifier)
-        return toWebExtensionError(nil, nil, @"cannot specify both windowId and tabId");
+        return makeUnexpected(@"it is illegal to specify both windowId and tabId");
```
```suggestion
        return makeUnexpected(@"it cannot specify both 'windowId' and 'tabId'");
```

xeenon replied:
My wording really requires a key to be passed to `toWebExtensionError` in the second param. An example from scripting:

```objc++
toErrorString(nil, @"details", @"it cannot specify both 'func' and 'function'. Please use 'func'")
```

If you can't give it a key, then your original wording is likely best. But I would say "invalid" instead of "illegal".

```suggestion
        return makeUnexpected(@"it is invalid to specify both 'windowId' and 'tabId'");
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -71,79 +74,128 @@
 {
     if (windowIdentifier && tabIdentifier)
-        return toWebExtensionError(nil, nil, @"cannot specify both windowId and tabId");
+        return makeUnexpected(@"it is illegal to specify both windowId and tabId");
```
```suggestion
        return makeUnexpected(@"it cannot specify both 'windowId' and 'tabId'");
```

xeenon replied:
Ditto from the other spot about a key or using "invalid" instead.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -322,9 +369,17 @@
 {
+    NSString *apiName;
+    // this method services sidePanel.getOptions and sidebarAction.getPanelPath
+    // we need to know which one we're servicing so we can return the correct API name in errors
```
```suggestion
```
The code speaks for itself here.

AriYoung00 committed:
Implement sidebarAction/sidePanel open/close/toggle API methods.
https://webkit.org/b/279684
rdar://135894111

Reviewed by Timothy Hatcher.

This PR fills in the implementation of the `sidebarAction.open()`,
`sidebarAction.close()`, `sidebarAction.toggle()`, and
`sidePanel.open()` APIs. It also fixes a bug in the implementation
of `sidebarAction.isOpen`. Finally, this PR also adds a reasonably
comprehensive set of unit tests that exercise the behavior of these
API methods.

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension _hasSidebar]): Replace WebExtension::hasSidebar with
WebExtension::hasAnySidebar
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::toExpected): Replace call to toWebExtensionError with
makeUnexpected
(WebKit::getSidebarWithIdentifiers): Replace calls to
toWebExtensionError with makeUnexpected
(WebKit::getOrCreateSidebarWithIdentifiers): Replace calls to
toWebExtensionError with makeUnexpected
(WebKit::scopedAPINameFor): Add helper method to create full API name
for methods which service both sidebarAction and sidePanel
(WebKit::WebExtensionContext::openSidebar): Replaced openSidebarForTab
with openSidebar (allows for less-convoluted code in other locations).
Removed call to fireActionClickEventIfNeeded, added call to delegate to
actually open sidebar.
(WebKit::WebExtensionContext::closeSidebar): Replaced closeSidebarForTab
with closeSidebar, added call to delegate to actually close sidebar.
(WebKit::WebExtensionContext::canProgrammaticallyOpenSidebar): Moved
this method to WebExtensionContext rather than WebExtensionSidebar,
added implementation.
(WebKit::WebExtensionContext::canProgrammaticallyCloseSidebar): Moved
this method to WebExtensionContext rather than WebExtensionSidebar,
added implementation.
(WebKit::WebExtensionContext::sidebarOpen): Simplified implementation
based on other changes in this PR, added condition that relevant tab
must have active user gesture to programmatically open sidebar.
(WebKit::WebExtensionContext::sidebarClose): Cleaned up implementation,
added condition that relevant tab must have active user gesture to
programmatically close sidebar.
(WebKit::WebExtensionContext::sidebarIsOpen): Removed code which checks
if window-specific sidebar is open (window-specific objects should never
directly be open).
(WebKit::WebExtensionContext::sidebarToggle): Fixed minor issues in
implementation, cleaned up implementation.
(WebKit::WebExtensionContext::sidebarGetTitle): Add API name to error
message
(WebKit::WebExtensionContext::sidebarSetTitle): Add API name to error
message
(WebKit::WebExtensionContext::sidebarGetOptions): Add API name to error
message
(WebKit::WebExtensionContext::sidebarSetOptions): Add API name to error
message
(WebKit::WebExtensionContext::openSidebarForTab): Deleted (renamed to
openSidebar).
(WebKit::WebExtensionContext::closeSidebarForTab): Deleted (renamed to
closeSidebar).
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasSidebarAction): Helper method to check whether
this extension uses sidebarAction
(WebKit::WebExtension::hasSidePanel): Helper method to check whether
this extension uses sidePanel
(WebKit::WebExtension::hasAnySidebar): Helper method to check whether
this extension uses sidebarAction or sidePanel
(WebKit::WebExtension::hasSidebar): Deleted (replaced with
hasAnySidebar).
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getOrCreateSidebar): Replace calls to
WebExtension::hasSidebar with WebExtension::hasAnySidebar
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(WebKit::WebExtensionSidebar::canProgrammaticallyOpenSidebar const): Deleted
(moved to WebExtensionContext).
(WebKit::WebExtensionSidebar::openSidebarWhenReady): Deleted.
(WebKit::WebExtensionSidebar::canProgrammaticallyCloseSidebar const): Deleted
(moved to WebExtensionContext).
(WebKit::WebExtensionSidebar::closeSidebarWhenReady): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Add declarations of
  hasSidebarAction, hasSidePanel, hasAnySidebar; remove declaration of
  hasSidebar
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Removed
  declarations of openSidebarForTab and closeSidebarForTab, added
declarations of openSidebar, closeSidebar,
canProgrammaticallyOpenSidebar, and canProgrammaticallyCloseSidebar
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Removed
  declarations of canProgrammaticallyOpenSidebar, openSidebarWhenReady,
canProgrammaticallyCloseSidebar, and closeSidebarWhenReady.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
(WebKit::parseWindowIdentifier): Fix bug (accidentally typo'd tabIdKey
rather than windowIdKey).
(WebKit::WebExtensionAPISidePanel::open): Add user gesture check
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::WebExtensionAPISidebarAction::open): Add user gesture check
(WebKit::WebExtensionAPISidebarAction::close): Add user gesture check
(WebKit::WebExtensionAPISidebarAction::toggle): Add user gesture check
(WebKit::WebExtensionAPISidebarAction::isOpen): Only return error for
invalid window ID if windowId has a value AND isValid(windowId) also
fails, rather than just if isValid(windowId) fails
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm: Add
  import for _WKWebExtensionSidebar.h, add default path to
sidePanelManifest global
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionClearTabPanel)): Renamed old test to fix typo
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionOpenFailsWithoutUserGesture)):
Added test ensuring that sidebarAction.open() fails if the current tab
has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionCloseFailsWithoutUserGesture)):
Added test ensuring that sidebarAction.close() fails if the current tab
has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionToggleFailsWithoutUserGesture)):
Added test ensuring that sidebarAction.toggle() fails if the current tab
has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionOpenSucceedsWithUserGesture)):
Added test ensuring that sidebarAction.open() succeeds if the current
tab has a user gesture. Also ensures that the relevant delegate method
is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionCloseSucceedsWithUserGesture)):
Added test ensuring that sidebarAction.close() succeeds if the current
tab has a user gesture. Also ensures that the relevant delegate method
is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarActionToggleSucceedsWithUserGesture)):
Added test ensuring that sidebarAction.toggle() succeeds if the current
tab has a user gesture. Also ensures that the relevant delegate method
is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpenForTabFailsWithoutUserGesture)):
Added test ensuring that sidePanel.open({ tabId: ... }) fails if the
current tab has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpenForWindowFailsWithoutUserGesture)):
Added test ensuring that sidePanel.open({ windowId: ... }) fails if the
current tab has no user gesture.
(TestWebKitAPI::TEST_F(WKWebExtnsionAPISidebar, SidePanelOpenForTabSucceedsWithUserGesture)):
Added test ensuring that sidePanel.open({ tabId: ... }) succeeds if the
current tab has a user gesture. Also ensures that the relevant delegate
method is called correctly.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpenForWindowSucceedsWithUserGesture)):
Added test ensuring that sidePanel.open({ windowId: ... }) succeeds if
the current tab has a user gesture. Also ensures that the relevant
delegate method is called correctly, and that the sidebar is opened in
the correct window.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidebarAcionClearTabPanel)): Deleted (renamed to SidebarActionClearTabPanel).
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h: Added
  presentSidebar, closeSidebar, and didUpdateSidebar callback
properties.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:presentSidebar:forExtensionContext:completionHandler:]):
Added.
(-[TestWebExtensionsDelegate _webExtensionController:closeSidebar:forExtensionContext:completionHandler:]):
Added.
(-[TestWebExtensionsDelegate _webExtensionController:didUpdateSidebar:forExtensionContext:]):
Added.

Canonical link: https://commits.webkit.org/283934@main
